[Woo POS] Coupons: Display coupon discount in cart – Dummy UI

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15334
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Adding a structure to display coupon discounts in the cart with a dummy UI. It's not yet clear how the UI will look, but I added the ability that we will need to have one way or another:

- Show discount on a coupon
- Show validation status on a coupon - invalid / validating

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Prerequisites: Enable `enableCouponsInPointOfSale` to `true`

### Valid Coupon

1. Open POS
2. Add Product
3. Add Valid Coupon
4. Check out
5. Confirm "Validating..." is shown on the coupon while the order is loading
6. Confirm discount is shown on the coupon when order is loaded

### Invalid Coupon

1. Open POS
2. Add Product
3. Add Valid Coupon
4. Check out
5. Confirm "Validating..." is shown on the coupon while the order is loading
6. Confirm "Invalid coupon" is shown on the coupon when order sync fails

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Additionally, tested that the changes do not affect cases without the feature flag.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/034b3505-6507-4c78-9e0c-c0655ff66236

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Define PointOfSaleCouponTotal for setting total coupon discount

staskus committed:
Add couponRowState calculation to CartViewHelper

staskus committed:
Set coupons totals in PointOfSaleOrderController

staskus committed:
Add dummy UI for displaying coupon state and discount in CouponRowView

staskus committed:
Update CartViewHelperTests for couponRowState

staskus committed:
Merge branch 'trunk' into task/15334-pos-coupons-display-coupon-discount-in-cart

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/ViewHelpers/CartViewHelperTests.swift:
```diff
@@ -85,6 +85,86 @@ struct CartViewHelperTests {
+        let coupon = CartCouponItem(id: UUID(), code: "TEST10")
+
+        // When, Then
+
```
```suggestion
```

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/CouponRowView.swift:
```diff
@@ -22,11 +24,29 @@ struct CouponRowView: View {
                 .frame(width: Constants.couponCardSize, height: Constants.couponCardSize)
 
-            VStack(alignment: .leading) {
+            VStack(alignment: .leading, spacing: Constants.itemTitleAndPriceSpacing * (1 / scale)) {
```
Should we move this calculation to a computed variable? Something like `dynamicSpacing` or similar?

iamgabrielma reviewed (approved):
Works well! Very nice ahead-planning! 🚢

staskus committed:
Update WooCommerce/WooCommerceTests/POS/ViewHelpers/CartViewHelperTests.swift

Co-authored-by: Gabriel Maldonado <gabriel.maldonado@automattic.com>

staskus committed:
Refactor spacing in CouponRowView for improved readability

- Introduced a dynamicSpacing property to calculate spacing based on scale.
- Updated VStack spacing to use dynamicSpacing instead of a direct calculation.
